File: intersection_y.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def get_count(head):
    c = 0
    while head:
        c += 1
        head = head.next
    return c

def get_intersection(head1, head2):
    c1 = get_count(head1)
    c2 = get_count(head2)
    d = abs(c1 - c2)
    longer = head1 if c1 > c2 else head2
    shorter = head2 if c1 > c2 else head1
    for _ in range(d):
        longer = longer.next
    while longer and shorter:
        if longer is shorter:
            return longer
        longer = longer.next
        shorter = shorter.next
    return None

if __name__ == "__main__":
    head1 = Node(3)
    head1.next = Node(6)
    head1.next.next = Node(9)
    head1.next.next.next = Node(15)
    head1.next.next.next.next = Node(30)

    head2 = Node(10)
    head2.next = head1.next.next.next  # intersection at 15

    res = get_intersection(head1, head2)
    if res:
        print("Intersection at node with data =", res.data)
    else:
        print("No intersection")

File: intersection_y.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node* next;
} Node;

Node* newNode(int d) {
    Node* n = (Node*)malloc(sizeof(Node));
    n->data = d;
    n->next = NULL;
    return n;
}

int getCount(Node* head) {
    int c = 0;
    while (head) { c++; head = head->next; }
    return c;
}

Node* getIntersection(Node* head1, Node* head2) {
    int c1 = getCount(head1);
    int c2 = getCount(head2);
    int d = abs(c1 - c2);
    Node *longer = (c1 > c2) ? head1 : head2;
    Node *shorter = (c1 > c2) ? head2 : head1;
    for (int i = 0; i < d; ++i) longer = longer->next;
    while (longer && shorter) {
        if (longer == shorter) return longer;
        longer = longer->next;
        shorter = shorter->next;
    }
    return NULL;
}

void printNode(Node* n) {
    if (n) printf("Intersection at node with data = %d\n", n->data);
    else printf("No intersection\n");
}

int main() {
    // Create two lists that intersect
    Node* head1 = newNode(3);
    head1->next = newNode(6);
    head1->next->next = newNode(9);
    head1->next->next->next = newNode(15);
    head1->next->next->next->next = newNode(30);

    Node* head2 = newNode(10);
    head2->next = head1->next->next->next; // points to node with 15

    Node* inter = getIntersection(head1, head2);
    printNode(inter);
    return 0;
}
